contrib/IECoreUSD/src/IECoreUSD/USDScene.cpp

```diff
@@ -0,0 +1,832 @@
+			return nullptr;
+		case SceneInterface::ThrowIfMissing :
+			throw IOException( "Child \"" + name.string() + "\" does not exist" );
+		case SceneInterface::CreateIfMissing :
+			throw InvalidArgumentException( "Child creation not supported" );
+		default:
+			return nullptr;
+	}
+}
+
+ConstSceneInterfacePtr USDScene::child( const SceneInterface::Name &name, SceneInterface::MissingBehaviour missingBehaviour ) const
+{
+	return const_cast<USDScene *>( this )->child( name, missingBehaviour );
+}
+
+SceneInterfacePtr USDScene::createChild( const SceneInterface::Name &name )
+{
+	throw IECore::NotImplementedException("USDScene::createChild not supported" );
+	return nullptr;
+}
+
+SceneInterfacePtr USDScene::scene( const SceneInterface::Path &path, SceneInterface::MissingBehaviour missingBehaviour )
+{
+	pxr::UsdPrim prim = m_root->root();
+
+	for( const Name &name : path )
+	{
+		prim = prim.GetChild( pxr::TfToken( name ) );
+	}
+	return new USDScene( m_root, prim );
+}
+
+ConstSceneInterfacePtr USDScene::scene( const SceneInterface::Path &path, SceneInterface::MissingBehaviour missingBehaviour ) const
+{
+	return const_cast<USDScene *>( this )->scene( path, missingBehaviour );
+}
+
+void USDScene::hash( SceneInterface::HashType hashType, double time, MurmurHash &h ) const
+{
+	SceneInterface::hash( hashType, time, h );
+
+	h.append( hashType );
+
+	auto transformHash = [this]( double time, MurmurHash &h )
+	{
+		if( pxr::UsdGeomXformable xformable = pxr::UsdGeomXformable( m_prim ) )
+		{
+			h.append( m_prim.GetPath().GetString() );
+			h.append( m_root->fileName() );
+
+			if( xformable.TransformMightBeTimeVarying() )
+			{
+				h.append( time );
+			}
+		}
+	};
+
+	auto boundsHash = [this]( double time, MurmurHash &h )
+	{
+		if( pxr::UsdGeomBoundable boundable = pxr::UsdGeomBoundable( m_prim ) )
+		{
+			h.append( m_prim.GetPath().GetString() );
+			h.append( m_root->fileName() );
+
+			if( boundable.GetExtentAttr().ValueMightBeTimeVarying() )
+			{
+				h.append( time );
+			}
+		}
+	};
+
+	auto objectHash = [this]( double time, MurmurHash &h )
+	{
+		if( isConvertable( m_prim ) )
+		{
+			h.append( m_prim.GetPath().GetString() );
+			h.append( m_root->fileName() );
+
+			if( isTimeVarying( m_prim ) )
+			{
+				h.append( time );
+			}
+		}
+	};
+
+	auto childNamesHash = [this]( double time, MurmurHash &h )
+	{
+		h.append( m_prim.GetPath().GetString() );
+		h.append( m_root->fileName() );
+		h.append( time );
+	};
+
+	auto hierarchyHash = [this]( double time, MurmurHash &h )
+	{
+		h.append( m_prim.GetPath().GetString() );
+		h.append( m_root->fileName() );
+		h.append( time );
+	};
+
+	switch( hashType )
+	{
+		case SceneInterface::TransformHash:
+			transformHash( time, h );
+			break;
+		case SceneInterface::AttributesHash:
+			break;
+		case SceneInterface::BoundHash:
+			boundsHash( time, h );
+			break;
+		case SceneInterface::ObjectHash:
+			objectHash( time, h );
+			break;
+		case SceneInterface::ChildNamesHash:
```

Comment:
I'm not sure how to implement ChildNamesHash & HierarchyHash given the USD API. I noticed there is a hashing  API in IECoreAlembic but I can't find similar functionality in USD.

Reply:
> Does USD really allow time-varying hierarchy?

I'll try and get an answer to this but in the meantime would you prefer faster but can't deal with changing hierarchy or more robust but slower?

Reply:
I want whatever we merge to give an accurate reflection of USD performance when scrubbing animation in Gaffer. So really I think we need to get an answer on this one. I have a vague feeling they might have a static hierarchy but some sort of active flag that can be animated?